### PR TITLE
Dendrite layer unit test

### PR DESCRIPTION
**Pull request by FishAres**
Added unit test for dendrite layer, k=1 k-winner activation for dendrites
Minor alterations in other files

### CODE REVIEW

File: nupic/research/frameworks/continuous_learning/dend_kwinners.py [hunk 1]
```diff
@@ -100,25 +91,26 @@ def __init__(
...
 class DendriteKWinners2dLocal(torch.autograd.Function):
+    """ x needs to be shaped as
+    (batch_size, no. units, dendrites per unit, 1)
+    """
+
     @staticmethod
     def forward(ctx, x, k):
 
-        batch_size, channels, h, w = x.shape
+        # batch_size, channels, h, w = x.shape
```
Comment (subutai) [inline nupic/research/frameworks/continuous_learning/dend_kwinners.py:108]: If we don't need this line, can we delete it?

File: nupic/research/frameworks/continuous_learning/dendrite_layers.py [hunk 2]
```diff
@@ -150,52 +142,39 @@ def __init__(self,
...
+    def forward(self, x, cat_projection=1.0):
         """ cat_proj here is an optional argument
         for a categorical "feedback" projection to
         the dendrite segments
         """
-        if self.act_fun_type is None:
+        if self.act_fun_type == "kwinner":
             return self.forward_kwinner(x, cat_projection)
         else:
             return self.forward_sigmoid(x, cat_projection)
```
Comment (subutai) [inline nupic/research/frameworks/continuous_learning/dendrite_layers.py:153]: Would recommend explicitly checking for "sigmoid" and then throwing an exception if it doesn't match "kwinner" or "sigmoid". This way a spelling mistake when calling it won't silently run as sigmoid.

File: nupic/research/frameworks/continuous_learning/utils.py [hunk 3]
```diff
@@ -108,59 +115,22 @@ def train_model(
...
-                               with the following parameters: model, batch_idx
-    :type pre_batch_callback: function
-    :param progress_bar: Optional :class:`tqdm` progress bar args.
-                         None for no progress bar
-    :type progress_bar: dict or None
-
-    :return: mean loss for epoch
-    :rtype: float
-    """
+
```
Comment (subutai) [inline nupic/research/frameworks/continuous_learning/utils.py:133]: Why remove the docstring? At the very least the new parameters should be documented. Not clear to me at all what they mean.

File: nupic/research/frameworks/continuous_learning/utils.py [hunk 4]
```diff
@@ -206,32 +174,40 @@ def train_model(
...
-        else:
-            loss.backward()
 
-        if freeze_params is not None:
-            freeze_fun(model, freeze_params, duty_cycles, freeze_pct)
+        loss.backward()
+
+        if freeze_modules is not None:
+            cnt = 0
+            # with torch.no_grad():
```
Comment (subutai) [inline nupic/research/frameworks/continuous_learning/utils.py:182]: Remove commented line if we don't need it.

File: nupic/research/frameworks/continuous_learning/utils.py [hunk 5]
```diff
@@ -206,32 +174,40 @@ def train_model(
...
+            freeze_output_layer(
+                model,
+                output_indices,
+                layer_type=layer_type,
+                linear_number=linear_number,
+            )
 
         t3 = time.time()
-        optimizer.step()
+        optimizer.step()  # step
```
Comment (subutai) [inline nupic/research/frameworks/continuous_learning/utils.py:196]: This comment is useless.

File: tests/unit/frameworks/continuous_learning/dendrite_layer_test.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+from nupic.research.frameworks.continuous_learning.dend_kwinners import (
+    DendriteKWinners2d,
+    DendriteKWinners2dLocal,
+)
+from nupic.research.frameworks.continuous_learning.dendrite_layers import DendriteOutput
+
+
+class DendKWinnerTest(unittest.TestCase):
+    def setUp(self):
+        x = torch.randn(3, 2, 4, 1)
```
Comment (subutai) [inline tests/unit/frameworks/continuous_learning/dendrite_layer_test.py:36]: Why not just self.x = torch.randn...?
